payments/paypal/paypal.go

```diff
@@ -0,0 +1,208 @@
+	rh := &paypalRequestHandler{
+		p: p,
+	}
+	var bp paypalBodyParams
+	bod, err := r.GetBody()
+	if err != nil {
+		return nil, err
+	}
+	err = json.NewDecoder(bod).Decode(&bp)
+	if err != nil {
+		return nil, err
+	}
+	rh.paymentID = bp.PaypalID
+	rh.userID = bp.PaypalUserID
+	if rh.paymentID == "" || rh.userID == "" {
+		return nil, errors.New("Payments requires a paypal_payment_id and paypal_user_id pair")
+	}
+
+	return rh, nil
+}
+
+func (r *paypalRequestHandler) Charge(amount uint64, currency string) (string, error) {
+	payment, err := r.p.client.GetPayment(r.paymentID)
+	if err != nil {
+		return "", err
+	}
+	if len(payment.Transactions) != 1 {
+		return "", fmt.Errorf("The paypal payment must have exactly 1 transaction, had %v", len(payment.Transactions))
+	}
+
+	if payment.Transactions[0].Amount == nil {
+		return "", fmt.Errorf("No amount in this transaction %v", payment.Transactions[0])
+	}
+
+	transactionValue := fmt.Sprintf("%.2f", float64(amount)/100)
+
+	if transactionValue != payment.Transactions[0].Amount.Total || payment.Transactions[0].Amount.Currency != currency {
+		return "", fmt.Errorf("The Amount in the transaction doesn't match the amount for the order: %v", payment.Transactions[0].Amount)
+	}
+
+	executeResult, err := r.p.client.ExecuteApprovedPayment(r.paymentID, r.userID)
+	if err != nil {
+		return "", err
+	}
+
+	return executeResult.ID, nil
+}
+
+func (p *paypalPaymentProvider) NewRefunder(ctx context.Context, r *http.Request) (payments.Refunder, error) {
+	// TODO ~ refund via paypal #58
+	return nil, errors.New("PayPal provider does not support refunds")
+}
+
+func (p *paypalPaymentProvider) NewPreauthorizer(ctx context.Context, r *http.Request) (payments.Preauthorizer, error) {
+	config := gcontext.GetConfig(ctx)
+	return &paypalRequestHandler{
+		siteURL: config.SiteURL,
+	}, nil
+}
+
+func (p *paypalPaymentProvider) getExperience() (*paypalsdk.WebProfile, error) {
+	p.profileMutex.Lock()
+	defer p.profileMutex.Unlock()
+
+	if p.profile != nil {
+		return p.profile, nil
+	}
+
+	experiences, err := p.client.GetWebProfiles()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed getting web profiles")
+	}
+
+	for _, profile := range experiences {
+		if profile.Name == "gocommerce" {
+			p.profile = &profile
+			return p.profile, nil
+		}
+	}
+
+	profile, err := p.client.CreateWebProfile(paypalsdk.WebProfile{
+		Name: "gocommerce",
+		InputFields: paypalsdk.InputFields{
+			NoShipping: 1,
+		},
+	})
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed creating web profile")
+	}
+
+	p.profile = profile
+	return profile, nil
+}
+
+func (r *paypalRequestHandler) Preauthorize(amount uint64, currency string, description string) (*payments.PreauthorizationResult, error) {
+	profile, err := r.p.getExperience()
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating paypal experience")
+	}
+
+	redirectURI := r.siteURL + "/gocommerce/paypal"
```

Comment:
Ok. It was a little unclear what those URLs actually pointed to. Is that a page we provide, or something they are required to implement? This would be good to document.